### PR TITLE
Implement oauth in KEB

### PR DESCRIPTION
**Pull request by polskikiel**
**Description**

Changes proposed in this pull request:

- Implemented oauth in KEB

**Related issue(s)**

### CODE REVIEW

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Authorization
+
+The Kyna Environment Broker provides the following ways to authorize users.
+
+## Basic auth
+
+To access the Kyma Environment Broker endpoints with basic authorization enabled, you must specify an Authorization Basic token header:
+
+```
+Authorization: Basic YnJva2VyOkVVMmpLQmVKOGc=
```
Comment (mszostok) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
Authorization: Basic {BASE64_ENCODED_CREDENTIALS}
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+```
+
+>**NOTE**: The basic auth implementation is currently being replaced by the Oauth2 implementation and will be deprecated soon.
+
+## Oauth2
+
+The Kyma Environment Broker allows to authorize users using the Oauth authentication. It is using the [ApiRule](https://github.com/kyma-project/kyma/blob/master/docs/api-gateway-v2/06-01-apirule.md) to provide a VirtualService and [Oathkeeper Access Rules](https://www.ory.sh/docs/oathkeeper/api-access-rules) according to the details specified in the CR.
```
Comment (mszostok) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
The Kyma Environment Broker allows to authorize users using the OAuth2 authorization. It is using the [ApiRule](https://github.com/kyma-project/kyma/blob/master/docs/api-gateway-v2/06-01-apirule.md) to provide a VirtualService and [Oathkeeper Access Rules](https://www.ory.sh/docs/oathkeeper/api-access-rules) according to the details specified in the CR.
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+```
+Authorization: Bearer {ACCESS_TOKEN}
+```
+
+### Access token
+
+Follow the tutorial to obtain a new access token.
+
+1. Create a OAuth2 Client:
```
Comment (mszostok) [inline docs/kyma-environment-broker/03-05-authorization.md]: add first step about exporting necessary env variables

---

which namespace should I use?

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  kind: OAuth2Client
+  metadata:
+    name: $CLIENT_NAME
+    namespace: $CLIENT_NAMESPACE
+  spec:
+    grantTypes:
+      - "client_credentials"
+    scope: "broker:write"
+    secretName: $CLIENT_NAME
+  EOF
```
Comment (mszostok) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
EOF
```

File: components/kyma-environment-broker/cmd/broker/main.go [hunk 5]
```diff
@@ -163,9 +157,22 @@ func main() {
...
+		Username: cfg.Auth.Username,
+		Password: cfg.Auth.Password,
+	}
+
+	// create and run broker OSB API in 2 modes:
+	// with basic auth
+	// with oauth
+	brokerBasicAPI := broker.New("", kymaEnvBroker, logger, &brokerCredentials)
+	brokerAPI := broker.New("/oauth", kymaEnvBroker, logger, nil)
+	n := negroni.New(negroni.NewRecovery())
```
Comment (mszostok) [inline components/kyma-environment-broker/cmd/broker/main.go]: with that solution I'm getting 404 when calling the https://kyma-env-broker.$domain/v2/catalog

and such response body:
```
404 page not found
Not Authorized
```

**expected:**
code: 401
body: "Not Authorized"

When I provide a problem basic then I got such reponse:
```
404 page not found
{"services":[{"id":"47c9dcbf-ff30-448e-ab36-d3bad66ba281","name":"kymaruntime","description":"[EXPERIMENTAL] Service Class for Kyma Runtime","bindable":true,"tags":["SAP","Kyma"],"plan_updateable":false,"plans":[{"id":"ca6e5357-707f-4565-bbbd-b3ab732597c6","name":"gcp","description":"GCP","metadata":{"displayName":"GCP"},"schemas":{"service_instance":{"create":{"parameters":{"$schema":"http://json-schema.org/draft-04/schema#","properties":{"autoScalerMax":{"type":"integer"},"autoScalerMin":{"type":"integer"},"components":{"items":{"enum":["Kiali","Jaeger"],"type":"string"},"type":"array"},"diskType":{"type":"string"},"machineType":{"enum":["n1-standard-2","n1-standard-4","n1-standard-8","n1-standard-16","n1-standard-32","n1-standard-64"],"type":"string"},"maxSurge":{"type":"integer"},"maxUnavailable":{"type":"integer"},"name":{"type":"string"},"region":{"enum":["asia-south1","asia-southeast1","asia-east2","asia-east1","asia-northeast1","asia-northeast2","australia-southeast1","europe-west2","europe-west1","europe-west4","europe-west6","europe-west3","europe-north1","us-west1","us-west2","us-central1","us-east1","us-east4","northamerica-northeast1","southamerica-east1"],"type":"string"},"volumeSizeGb":{"type":"integer"},"zone":{"enum":["asia-east1-a","asia-east1-b","asia-east1-c","asia-east2-a","asia-east2-b","asia-east2-c","asia-northeast1-a","asia-northeast1-b","asia-northeast1-c","asia-northeast2-a","asia-northeast2-b","asia-northeast2-c","asia-south1-a","asia-south1-b","asia-south1-c","asia-southeast1-a","asia-southeast1-b","asia-southeast1-c","australia-southeast1-a","australia-southeast1-b","australia-southeast1-c","europe-north1-a","europe-north1-c","europe-north1-b","europe-west1-b","europe-west1-c","europe-west1-d","europe-west2-a","europe-west2-b","europe-west2-c","europe-west3-a","europe-west3-b","europe-west3-c","europe-west4-a","europe-west4-b","europe-west4-c","europe-west6-a","europe-west6-b","europe-west6-c","northamerica-northeast1-a","northamerica-northeast1-b","northamerica-northeast1-c","southamerica-east1-a","southamerica-east1-b","southamerica-east1-c","us-central1-a","us-central1-b","us-central1-c","us-central1-f","us-east1-b","us-east1-c","us-east1-d","us-east4-a","us-east4-b","us-east4-c","us-west1-a","us-west1-b","us-west1-c","us-west2-a","us-west2-b","us-west2-c"],"type":"string"}},"required":["name"],"type":"object"}},"update":{"parameters":null}},"service_binding":{"create":{"parameters":null}}}},{"id":"4deee563-e5ec-4731-b9b1-53b42d855f0c","name":"azure","description":"Azure","metadata":{"displayName":"Azure"},"schemas":{"service_instance":{"create":{"parameters":{"$schema":"http://json-schema.org/draft-04/schema#","properties":{"autoScalerMax":{"type":"integer"},"autoScalerMin":{"type":"integer"},"components":{"items":{"enum":["Jaeger","Kiali"],"type":"string"},"type":"array"},"diskType":{"type":"string"},"machineType":{"enum":["Standard_D8_v3"],"type":"string"},"maxSurge":{"type":"integer"},"maxUnavailable":{"type":"integer"},"name":{"type":"string"},"region":{"enum":["westeurope","eastus","eastus2","centralus","northeurope","southeastasia","japaneast","westus2","uksouth"],"type":"string"},"volumeSizeGb":{"minimum":50,"type":"integer"},"zone":{"type":"string"}},"required":["name"],"type":"object"}},"update":{"parameters":null}},"service_binding":{"create":{"parameters":null}}}}],"metadata":{"displayName":"Kyma Runtime","documentationUrl":"kyma-project.io","longDescription":"Kyma Runtime experimental service class","providerDisplayName":"SAP"}}]}
```

with code 404 and content type `text/plain; charset=utf-8`


---

oauth endpoint works properly.

File: components/kyma-environment-broker/internal/broker/server.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	Password string
+}
+
+// copied from github.com/pivotal-cf/brokerapi/api.go
+func New(serviceBroker domain.ServiceBroker, logger lager.Logger, brokerCredentials *BrokerCredentials) *mux.Router {
+	router := mux.NewRouter()
+
+	AttachRoutes(router, serviceBroker, logger)
+
+	if brokerCredentials != nil {
```
Comment (mszostok) [inline components/kyma-environment-broker/internal/broker/server.go]: the middlewares are added on stack and right now you are changing the default behavior the `router.Use(middlewares.AddCorrelationIDToContext)` should be before the auth

File: components/kyma-environment-broker/cmd/broker/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/kyma-environment-broker/cmd/broker/main.go]: It finally worked using the ServeMux

Comment (polskikiel) [reply components/kyma-environment-broker/cmd/broker/main.go]: Fixed

File: components/kyma-environment-broker/internal/broker/server.go (hunk 6 above)
Comment (polskikiel) [reply components/kyma-environment-broker/internal/broker/server.go]: Fixed

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+## Basic auth
+
+To access the Kyma Environment Broker endpoints with basic authorization enabled, you must specify an Authorization Basic token header:
+
+```
+Authorization: Basic {BASE64_ENCODED_CREDENTIALS}
+```
+
+>**NOTE**: The basic auth implementation is currently being replaced by the Oauth2 implementation and will be deprecated soon.
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
>**NOTE**: This implementation is currently being replaced by the Oauth2 implementation and will be deprecated soon.
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
+# Authorization
+
+The Kyna Environment Broker provides the following ways to authorize users.
+
+## Basic auth
+
+To access the Kyma Environment Broker endpoints with basic authorization enabled, you must specify an Authorization Basic token header:
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
To access the Kyma Environment Broker endpoints with the Basic authorization enabled, specify the `Authorization: Basic` token header:
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+```
+Authorization: Basic {BASE64_ENCODED_CREDENTIALS}
+```
+
+>**NOTE**: The basic auth implementation is currently being replaced by the Oauth2 implementation and will be deprecated soon.
+
+## Oauth2
+
+The Kyma Environment Broker allows to authorize users using the OAuth2 authorization. It is using the [ApiRule](https://github.com/kyma-project/kyma/blob/master/docs/api-gateway-v2/06-01-apirule.md) to provide a VirtualService and [Oathkeeper Access Rules](https://www.ory.sh/docs/oathkeeper/api-access-rules) according to the details specified in the CR.
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
Kyma Environment Broker allows for authorizing users with the OAuth2 authorization. It uses the [ApiRule](https://github.com/kyma-project/kyma/blob/master/docs/api-gateway-v2/06-01-apirule.md) to provide a VirtualService and [Oathkeeper Access Rules](https://www.ory.sh/docs/oathkeeper/api-access-rules) according to the details specified in the CR.
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+To authorize with the Kyma Environment Broker, use an OAuth2 client registered through the Hydra Maester controller. Look for more information [here](https://github.com/kyma-project/kyma/blob/master/docs/api-gateway-v2/08-01-exposesecure.md#register-an-oauth2-client-and-get-tokens).
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
To authorize with Kyma Environment Broker, use an OAuth2 client registered through the Hydra Maester controller. For more information, read [this](https://github.com/kyma-project/kyma/blob/master/docs/api-gateway-v2/08-01-exposesecure.md#register-an-oauth2-client-and-get-tokens) document.
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+To access the Kyma Environment Broker endpoints with OAuth2 authorization enabled, use the `/oauth` prefix. For example:
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
To access the Kyma Environment Broker endpoints with the OAuth2 authorization enabled, use the `/oauth` prefix. For example:
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+```
+/oauth/v2/catalog
+```
+
+You must also specify an Authorization Bearer token header:
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
You must also specify the `Authorization: Bearer` token header:
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+```
+Authorization: Bearer {ACCESS_TOKEN}
+```
+
+### Access token
+
+Follow the tutorial to obtain a new access token.
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
Follow these steps to obtain a new access token:
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    export CLIENT_NAMESPACE={YOUR_CLIENT_NAMESPACE}
+    ```
+
+  - The domain of your cluster:
+
+    ```shell
+    export DOMAIN={CLUSTER_DOMAIN}
+    ```
+
+2. Create a OAuth2 Client:
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
2. Create an OAuth2 client:
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+```
+
+>**NOTE**: This implementation is currently being replaced by the Oauth2 implementation and will be deprecated soon.
+
+## OAuth2 authorization
+
+The Kyma Environment Broker allows to authorize users using the OAuth2 authorization. It is using the [ApiRule](https://github.com/kyma-project/kyma/blob/master/docs/api-gateway-v2/06-01-apirule.md) to provide a [VirtualService](https://istio.io/docs/reference/config/networking/virtual-service/) and [Oathkeeper Access Rules](https://www.ory.sh/docs/oathkeeper/api-access-rules) according to the details specified in the CR.
+To authorize with the Kyma Environment Broker, use an OAuth2 client registered through the [Hydra Maester controller](https://github.com/ory/k8s/blob/master/docs/helm/hydra-maester.md).
+
+To access the Kyma Environment Broker endpoints with the OAuth2 authorization enabled, use the `/oauth` prefix. For example:
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
To access the Kyma Environment Broker endpoints with the OAuth2 authorization enabled, use the `/oauth` prefix. For example:
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+```
+/oauth/v2/catalog
+```
+
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
>**NOTE:** If you do not use the `/oauth` prefix, the Basic authorization is performed.
```

File: docs/kyma-environment-broker/03-05-authorization.md [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+You must also specify the `Authorization: Bearer` token header:
+
+```
+Authorization: Bearer {ACCESS_TOKEN}
+```
+
+### Access token
```
Comment (klaudiagrz) [inline docs/kyma-environment-broker/03-05-authorization.md]:
```suggestion
### Get the access token
```

**Pull request merged**